sql/2020/07_SEO/pages_canonicals_by_device.sql

```diff
@@ -0,0 +1,128 @@
+#standardSQL
+# page canonicals metrics grouped by device
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+# returns all the data we need from _canonicals
+CREATE TEMPORARY FUNCTION get_canonicals_info(canonicals_string STRING)
+RETURNS STRUCT<
+
+  ##Antoine
+  canonicals ARRAY<STRING>, 
+  has_self_canonical BOOL, 
+  is_canonicalized BOOL,
+  http_canonicals ARRAY<STRING>, 
+  has_canonical_mismatch BOOL, 
+  canonical_raw ARRAY<STRING>, 
+  canonical_rendered ARRAY<STRING>
+
+
+
+  # add more properties here...
+
+> LANGUAGE js AS '''
+var result = {};
+try {
+    //var canonicals = JSON.parse(canonicals_string); // LIVE
+
+    // TEST
+    // Fill with test data 
```

Comment:
I created this test object to help fix bugs and simulate data:

```
    // TEST
    var wpt_bodies = {
      "canonicals": {
          "rendered": {
              "html_link_canoncials": [
                  "https://btpi.com/"
              ]
          },
          "raw": {
              "html_link_canoncials": [
                  "https://btpi.com/"
              ]
          },
          "self_canonical": Math.floor(Math.random() * 2) == 0,
          "other_canonical": Math.floor(Math.random() * 10) == 0,
          "canonicals": [
              "https://btpi.com/"
          ],
          "url": "https://btpi.com/",
          "http_header_link_canoncials": [],
          "canonical_missmatch": Math.floor(Math.random() * 20) == 0
      }
    }
    if (Math.floor(Math.random() * 50) == 0) {
        wpt_bodies.canonicals.canonicals = []; // sometimes no canonicals
    }
    if (Math.floor(Math.random() * 30) == 0) {
        wpt_bodies.canonicals.raw = {}; // sometimes no raw canonicals
    }
    if (Math.floor(Math.random() * 25) == 0) {
        wpt_bodies.canonicals.rendered.html_link_canoncials = ["https://someoneelse.com/"]; // sometimes rendering changes it
    }
    if (Math.floor(Math.random() * 60) == 0) {
        wpt_bodies.canonicals.http_header_link_canoncials = ["https://someoneelse.com/"]; // sometimes header exists
    }
```

Reply:
Added to the new commit that I'll push with your modifications applied :)